project/src/com/google/daggerquery/executor/models/Query.java

```diff
@@ -79,20 +86,71 @@ public Query(String typeName, String... parameters) {
   /**
    * Executes query on a {@link BindingGraph}.
    *
-   * <p>For all queries return a list with string. Strings content depends on a query name.
-   * <p>For `deps` queries each string represent exactly one dependency.
+   * <p>For all queries return a list with strings. Strings content depends on a query name.
+   *
+   * <ul>
+   * <li>For `deps` query each string represent exactly one dependency.
+   * <li>For `allpaths` query each string contains a path between {@code source} and {@code target} nodes.
+   * The connection between nodes is shown with construction '->'. For example, one of the possible paths
+   * may look like this: "com.google.First -> com.google.Second -> com.google.Third".
+   * </ul>
    */
   public List<String> execute(BindingGraph bindingGraph) {
-    if (name.equals("deps")) {
-      String sourceNode = parameters[0];
-      if (!bindingGraph.getAdjacencyListMap().containsKey(sourceNode)) {
-        throw new IllegalArgumentException("Specified source node doesn't exist.");
-      }
+    String source = parameters[0];
 
-      return bindingGraph.getAdjacencyListMap().get(sourceNode).getDependencyList()
-          .stream().map(DependencyProto.Dependency::getTarget).sorted().collect(toList());
+    if (!bindingGraph.getAdjacencyListMap().containsKey(source)) {
+      throw new IllegalArgumentException("Specified source node doesn't exist.");
+    }
+
+    switch (name) {
+      case DEPS_QUERY_NAME:
+        return bindingGraph.getAdjacencyListMap().get(source).getDependencyList()
+            .stream().map(Dependency::getTarget).sorted().collect(toList());
+
+      case ALLPATHS_QUERY_NAME:
+        String target = parameters[1];
+        Set<String> visitedNodes = new HashSet<>();
+        List<List<String>> result = new ArrayList<>();
+        List<String> path = new ArrayList<>();
+
+        findAllPaths(source, target, path, bindingGraph, visitedNodes, result);
+        return result.stream().map(list -> String.join(" -> ", list)).collect(toList());
     }
 
     throw new NotImplementedException();
   }
+
+  /**
+   * Traverses a {@link BindingGraph} starting from {@code source} node.
+   *
+   * <p>Puts all processed nodes in a {@code visitedNodes} set to avoid loops. Constructs a {@code path} which is
+   * an instance of {@link List<String>}. At each recursive level this variable contains a correct path in a graph from
+   * some root node which was passed in the first call and {@code target} node which is the same for all calls.
```

Comment:
Do you mean the *previous* call rather than the *first* call? (which I interpret it to be the `source` node tied to the query) :stuck_out_tongue:

Reply:
Sorry you're right, I misinterpreted the code.